projects/panoptic/index.js

```diff
@@ -0,0 +1,247 @@
+const sdk = require('@defillama/sdk')
+const { getLogs } = require('../helper/cache/getLogs')
+
+const FACTORY = '0x000000000000010a1DEc6c46371A28A071F8bb01'
+const SFPM = '0x0000000000000DEdEDdD16227aA3D836C5753194'
+const startBlocks = {
+  ethereum: 21389983,
+}
+
+const Q96 = BigInt(2) ** BigInt(96)
+
+function chainTvl(chain) {
+  return async (api) => {
+    const START_BLOCK = startBlocks[chain]
+    const poolDeployedLogs = (
+      await getLogs({
+        api,
+        target: FACTORY,
+        fromBlock: START_BLOCK,
+        topic: 'PoolDeployed(address,address,address,address)',
+      })
+    )
+
+    // can contain SFPM pools that don't have options markets yet
+    const poolInitializedLogs = (
+      await getLogs({
+        api,
+        target: SFPM,
+        fromBlock: START_BLOCK,
+        topic: 'PoolInitialized(address,uint64,int24,int24)',
+      })
+    )
+
+    const block = api.block
+
+    const poolData = {}
+
+    for (let log of poolDeployedLogs) 
+      poolData[`0x${log.topics[2].substr(-40)}`.toLowerCase()] = {marketAddress: `0x${log.topics[1].substr(-40)}`.toLowerCase()}
+    
+
+    for (let log of poolInitializedLogs) {
+      const V3PoolAddress = `0x${log.topics[1].substr(-40)}`.toLowerCase()
+      if (!poolData?.[V3PoolAddress]) poolData[V3PoolAddress] = {marketAddress: '0x3327b4D450fbB3a4b780510489C259D85776D559'.toLowerCase()} // random address
+    }
+
+    const token0Calls = Object.keys(poolData).map((V3Pool) => ({ target: V3Pool }))
+    const token1Calls = Object.keys(poolData).map((V3Pool) => ({ target: V3Pool }))
+
+    const token0Results = await sdk.api.abi.multiCall({
+      abi: "function token0() view returns (address)",
+      calls: token0Calls,
+      block,
+      chain,
+    })
+    
+    token0Results.output.forEach((call, i) => {
+      poolData[call.input.target].token0 = call.output
+    })
+
+    const token1Results = await sdk.api.abi.multiCall({
+      abi: "function token1() view returns (address)",
+      calls: token1Calls,
+      block,
+      chain,
+    })
+    
+    token1Results.output.forEach((call, i) => {
+      poolData[call.input.target].token1 = call.output
+    })
+
+    // Create balance calls for both tokens of each market
+    const balanceCalls = []
+
+    const LiquidityOwnedTokens = []
+    
+    // Iterate through markets array directly
+    for (let [V3Pool, poolDataEntry] of Object.entries(poolData)) {
+      const mintLogs = (
+        await getLogs({
+          api,
+          target: V3Pool,
+          fromBlock: START_BLOCK,
+          extraKey: "mintCache",
```

Comment:
would be nice if the topics were automatically included as part of the cache key -- otherwised cached logs from the same address that don't match the topics will fail ABI decoding and cause errors

Reply:
yeah, you are right, it was a oversight on my side. Issue is, we cant fix it now without losing all current cache.

Reply:
I am worried that this implementation is not scalable, now we are going to track all mint and burn logs in a cache file, over time we will hit the file size limit :(

Reply:
Got it -- I had a feeling you would say something along those lines.

Unfortunately, it is impossible to reliably pull a list of liquidity positions without indexing on every mint and burn event at some level of the stack; thus, our options are to either find a way to retrieve those events that scales with your architecture, or pull indexed data from our subgraph indexer.

Happy to implement the latter of the two options if you make an exception to the rule about outside indexers in the README.

Reply:
yeah, this would have been a lot easier if there uni v3 NFTs were minted and kept, but I am guessing that you avoiding that to save gas?


Yeah, we can make an exception and use subgraph in your case

Reply:
*we have helper code that automatically unwrap the nfts and pull token value

Reply:
The positions are represented as NFTs, but they are ERC-1155 tokens from our custom position manager. It is possible to compute the TVL given a list of ids and supplies, but retrieving that information is impossible without indexing on 0.25-1x the amount of events we currently look at somewhere in the stack.